Move LCA calculation to the codebase where possible

pchiusano opened the pull request:
Lowest common ancestor calculations are used for computing merge results. This PR moves the lowest common ancestor calculation into the codebase where possible, so it's computed by SQL without having to deserialize any of the branches. This should be a lot faster.

Before this change, merging two branches with unrelated histories or with a common ancestor far in the past would require deserializing all the intermediate points in history back to their common ancestor.

To use, there's a new `Command`:

```Haskell
  -- Merge two branches, using the codebase for the LCA calculation where possible.
  Merge :: Branch.MergeMode -> Branch m -> Branch m -> Command m i v (Branch m)
```

There's one loose end: #1958. Until #1958, `push` will probably still be slow.

@aryairani did all the SQL wizardry and I just wrote some of the plumbing code.

@runarorama you can give this a shot on your copy of base that you're working on.

pchiusano committed:
Plumbing code to support LCA calculation in the codebase

pchiusano commented on parser-typechecker/src/Unison/Codebase/Branch.hs:
```diff
@@ -380,33 +381,42 @@ discardHistory0 = over children (fmap tweak) where
-merge' mode b1 b2 | isEmpty b2 = case mode of
+merge' = merge'' lca
+
+merge'' :: forall m . Monad m
+        => (Branch m -> Branch m -> m (Maybe (Branch m))) -- lca calculator
+        -> MergeMode
+        -> Branch m
+        -> Branch m
+        -> m (Branch m)
```
The `Branch.merge` implementation is now parameterized on a monadic function for calculating the LCA, which could be as simple as `Branch.lca` (for the in memory LCA calculation).

pchiusano commented on parser-typechecker/src/Unison/Codebase/Branch.hs:
```diff
@@ -380,33 +381,42 @@ discardHistory0 = over children (fmap tweak) where
+merge'' lca mode (Branch x) (Branch y) =
   Branch <$> case mode of
-               RegularMerge -> Causal.threeWayMerge combine x y
+               RegularMerge -> Causal.threeWayMerge' lca' combine x y
```
`Causal.threeWayMerge` also picks up the same parameter - how to calculate the LCA.

pchiusano commented on parser-typechecker/src/Unison/Codebase/Editor/Command.hs:
```diff
@@ -173,6 +173,9 @@ data Command m i v a where
   -- Like `LoadLocalRootBranch`.
   LoadLocalBranch :: Branch.Hash -> Command m i v (Branch m)
 
+  -- Merge two branches, using the codebase for the LCA calculation where possible.
+  Merge :: Branch m -> Branch m -> Command m i v (Branch m)
```
Within `HandleInput`, `eval $ Merge b1 b2` can be used in place of calling `Branch.merge b1 b2` directly. I didn't go through `HandleInput` and do all these replacements though.

aryairani replied:
Ok

pchiusano commented on parser-typechecker/src/Unison/Codebase/FileCodebase.hs:
```diff
@@ -188,6 +189,7 @@ codebase1' syncToDirectory branchCache fmtV@(S.Format getV putV) fmtA@(S.Format
     getTermsOfType r = listDirAsReferents (typeIndexDir path r)
     getTermsMentioningType :: Reference -> m (Set Referent.Id)
     getTermsMentioningType r = listDirAsReferents (typeMentionsIndexDir path r)
+    lca b1 b2 = Branch.lca b1 b2
```
File codebase just uses in memory LCA as before.

pchiusano commented on parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs:
```diff
@@ -685,6 +685,20 @@ sqliteCodebase root = do
+            else Branch.lca b1 b2
+            where
+              sqlLca :: Branch.Hash -> Branch.Hash -> m (Maybe Branch.Hash)
+              sqlLca h1 h2 = undefined h1 h2 -- insert sql wizardry
```
@aryairani you can fill in the SQL wizardry here

aryairani replied:
Done.

pchiusano commented on parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs:
```diff
@@ -685,6 +685,20 @@ sqliteCodebase root = do
+          lca b1 b2 = do
+            eb1 <- isCausalHash (Branch.headHash b1)
+            eb2 <- isCausalHash (Branch.headHash b2)
+            if eb1 && eb2 && False {- todo, remove this -} then do
```
We check that the two hashes exist in the codebase before trying the SQL version.

The SQL wizardry is switched off by this `&& False`, but once the SQL wizardry actually does something other than crash, comment out this `&& False`.

aryairani replied:
I removed the `&& False`.

aryairani committed:
add sql wizardry

pchiusano committed:
Add calls to Merge command in place of using Branch.merge directly

currently failing:
stack exec transcripts -- merge.md

giving different results:
stack exec transcripts -- delete.md
stack exec transcripts -- squash.md

pchiusano committed:
Add term, type, and decl caching

pchiusano committed:
Merge pull request #1952 from unisonweb/feature/codebase-lca-caching

Add term, type, and decl caching to the new codebase format

aryairani committed:
fix Sqlite.Queries.lca algorithm

pchiusano committed:
Merge pull request #1955 from unisonweb/fix/codebase-lca-wrapper

fix Sqlite.Queries.lca algorithm

aryairani reviewed (approved):
Go team!
